### PR TITLE
ng_ipv6_addr: initial import

### PR DESCRIPTION
**Pull request by authmillenon**
The motivation behind this module is to allow for address handling functions where they, but not the full IPv6 functionality might be needed. This new version of the IPv6 address data type utilizes the byteorder.h header to enforce network byte-order on its members.

### CODE REVIEW

**Comment by authmillenon** [comment]
Rebased

File: sys/include/net/ng_ipv6/addr.h [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+ *      </a>
+ *
+ * @param[in] addr  An IPv6 address.
+ *
+ * @return  true, if @p addr is loopback address,
+ * @return  false, otherwise.
+ */
+static inline bool ng_ipv6_addr_is_loopback(const ng_ipv6_addr_t *addr)
+{
+    return addr->u64[0].u64 == 0 &&
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: parens

File: sys/include/net/ng_ipv6/addr.h [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+ *      </a>
+ *
+ * @param[in] addr  An IPv6 address.
+ *
+ * @return  true, if @p addr is link-local address,
+ * @return  false, otherwise.
+ */
+static inline bool ng_ipv6_addr_is_link_local(const ng_ipv6_addr_t *addr)
+{
+    return (byteorder_ntohll(addr->u64[0]) == 0xfe80000000000000) ||
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: Can I haz a constant for this prefix?

File: sys/include/net/ng_ipv6/addr.h [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+ *      </a>
+ *
+ * @param[in] addr  An IPv6 address.
+ *
+ * @return  true, if @p addr is solicited-node multicast address,
+ * @return  false, otherwise.
+ */
+static inline bool ng_ipv6_addr_is_solicited_node(const ng_ipv6_addr_t *addr)
+{
+    return (byteorder_ntohll(addr->u64[0]) == 0xff02000000000000) &&
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: see above (or do I miss a good reason against that?)

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+            0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f
+        }
+    };
+    ng_ipv6_addr_t b = { {
+            0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
+            0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f
+        }
+    };
+    TEST_ASSERT_EQUAL_INT(false, ng_ipv6_addr_equal(&a, &b));
```
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Try with different IP addresses. E.g. with addresses that only differs at the first, the last or an arbitrary bit or differs only for two bits.

File: sys/include/net/ng_ipv6/addr.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/ng_ipv6/addr.h]: (also applies for https://github.com/RIOT-OS/RIOT/pull/2433#discussion_r24830428) I don't quite know what other use-case would be there for such a - then - public, other than using it here. The macros basically would be something like `NG_IPV6_ADDR_LINK_LOCAL_PREFIX_UINT64`. Then why not have a `NG_IPV6_ADDR_LINK_LOCAL_PREFIX_UINT32`, `NG_IPV6_ADDR_LINK_LOCAL_PREFIX_UINT16`, or `NG_IPV6_ADDR_LINK_LOCAL_PREFIX_UINT8`? Additionally for solicited node addresses the multicast address is prefixed with `0xff02000000000000` in this special case of the solicited node address. Link-local multicast addresses are generally prefixed with just `ff02::/4`, which would mean either more comparison or more weird public macros.

TL;DR: Too many special cases to provide them publically.

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c (hunk 4 above)
Comment (authmillenon) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Done. See below.

File: sys/include/net/ng_ipv6/addr.h (hunk 3 above)
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: I just don't like literals. And I don't see a real problem with global macros (one could mark and document them as internal). But maybe you're right and in this case it's not 100% necessary - it just hurts my feeling of style. :wink:

**Comment by OlegHahm** [comment]
I see your sneaky plan to make me spend more beer for overlooking a PR that needs squashing. ;)

**Comment by authmillenon** [comment]
True! It absolutely has nothing to do with me not feeling responsible to set this Label ;P

**Comment by authmillenon** [comment]
Reviewers, ping?

**Comment by OlegHahm** [comment]
As usual: I have to motivate myself to look at the unit tests again. ;)

**Comment by LudwigOrtmann** [comment]
On the samr21-xpro:
```
$ make clean tests-ipv6_addr flash
...
$ make term
...
2015-02-19 17:38:33,325 - INFO # kernel_init(): This is RIOT! (Version: 2014.12-451-gd6680-ng_ipv6_addr/feat/initial)
2015-02-19 17:38:33,330 - INFO # kernel_init(): jumping into first task...
2015-02-19 17:38:33,345 - INFO # ..............................................
2015-02-19 17:38:33,346 - INFO # OK (46 tests)
```

**Comment by OlegHahm** [comment]
I'm not talking about running the tests btw ;)

**Comment by LudwigOrtmann** [comment]
I understood that, just wanted to report this.

File: sys/include/net/ng_ipv6/addr.h [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+                                               0, 0, 0, 0, 0, 0, 0, 2 }}
+
+/**
+ * @brief   Values for the flag field in multicast addresses.
+ *
+ * @see <a href="http://tools.ietf.org/html/rfc4291#section-2.7">
+ *          RFC 4291, section 2.7
+ *      </a>
+ */
+typedef enum {
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: `__attribute__((packed))`?

File: sys/include/net/ng_ipv6/addr.h [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+} ng_ipv6_addr_mcast_flag_t;
+
+/**
+ * @brief   Values for the scope field in multicast addresses.
+ *
+ * @see <a href="http://tools.ietf.org/html/rfc4291#section-2.7">
+ *          RFC 4291, section 2.7
+ *      </a>
+ */
+typedef enum {
```
Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: `__attribute__((packed))`?

Comment (authmillenon) [inline sys/include/net/ng_ipv6/addr.h]: Didn't we reallized that this does weird stuff on some platforms?

Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: I thought that was only for packed structs. @thomaseichinger, do you recall?

Comment (gebart) [inline sys/include/net/ng_ipv6/addr.h]: There is also an `-fshort-enums` in gcc and clang.

Comment (OlegHahm) [inline sys/include/net/ng_ipv6/addr.h]: That would be even better - because it's much easier to disable it then in case of a problem.

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c (hunk 4 above)
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: superb

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c [hunk 7]
```diff
@@ -0,0 +1,687 @@
...
+    TEST_ASSERT_EQUAL_INT(0, a.u64[0].u64); /* Don't trust the macro ;) */
+    TEST_ASSERT_EQUAL_INT(0, a.u64[1].u64);
+    TEST_ASSERT_EQUAL_INT(true, ng_ipv6_addr_is_unspecified(&a));
+}
+
+static void test_ipv6_addr_is_multicast_not_multicast(void)
+{
+    ng_ipv6_addr_t a = { {
+            0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+            0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f
```
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Maybe to check for byteorder make it:
```
0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x01, 0xff
```

Comment (authmillenon) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Shouldn't that be

```
0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x01, 0xff,
0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f
```

And aren't byte order tests the task of that module, not the one that uses it?

Comment (OlegHahm) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Yes, your version is correct. And yes, it's not the task of this module to check for byte order, but if we can check for two things at the same time without any drawback, why not?

Comment (authmillenon) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Done

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c [hunk 8]
```diff
@@ -0,0 +1,687 @@
...
+
+static void test_ipv6_addr_init_prefix(void)
+{
+    ng_ipv6_addr_t a, b = { {
+            0x00, 0x01, 0x02, 0x02, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+        }
+    };
+
+    ng_ipv6_addr_init_prefix(&a, &b, 31);
+    TEST_ASSERT_EQUAL_INT(true, ng_ipv6_addr_equal(&a, &b));
```
Comment (OlegHahm) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: I don't get this test. Wouldn't it make more sense if a and b would be initially different?

**Comment by OlegHahm** [comment]
ACK apart from https://github.com/RIOT-OS/RIOT/pull/2433#discussion_r25075648

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c (hunk 8 above)
Comment (authmillenon) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: They are, or more correct: `a`'s value is unknown, since it is not initialized.

Comment (OlegHahm) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: Aaah / read `, ` for a `= `.

**Comment by OlegHahm** [comment]
Squash and go then (if Travis is happy).

File: tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c (hunk 8 above)
Comment (authmillenon) [inline tests/unittests/tests-ipv6_addr/tests-ipv6_addr.c]: This is C, not Python (or is this possible oO?)

**Comment by authmillenon** [comment]
Squashed; waiting for travis
